# Gaussian Elimination — Apples, Bananas, and Carrots 🍎🍌🥕

This notebook demonstrates how to solve a **system of linear equations** using **Gaussian elimination** — one of the most fundamental algorithms in linear algebra.

## 🎯 Learning Objectives
- Understand the systematic approach to solving linear systems
- Learn the two phases: **Forward Elimination** and **Back Substitution**
- See how elimination transforms a system into **upper triangular form**
- Connect the geometric interpretation with algebraic steps
- Understand why this method is computationally efficient

## 🔑 Key Concepts
1. **Linear System**: Multiple equations with multiple unknowns
2. **Augmented Matrix**: Combining coefficient matrix with constants
3. **Row Operations**: The three elementary operations that preserve solutions
4. **Upper Triangular Form**: Simplified structure enabling easy solution
5. **Back Substitution**: Working backwards from the last equation

## 🌟 Why This Matters
Gaussian elimination is the backbone of:
- **Linear algebra software** (NumPy, MATLAB, etc.)
- **Machine learning** (least squares, neural network training)
- **Computer graphics** (transformations, projections)
- **Engineering simulations** (finite element analysis)
- **Economics** (input-output models)

---

In [1]:
import numpy as np
import matplotlib.pyplot as plt

## 1. Problem Setup: Real-World Linear System

**Scenario**: You're at a farmer's market buying fruit in different combinations:

**Purchase 1**: 1 apple + 1 banana + 3 carrots = €15  
**Purchase 2**: 1 apple + 2 bananas + 4 carrots = €21  
**Purchase 3**: 2 apples + 1 banana + 1 carrot = €14  

**Question**: What's the individual price of each fruit?

### Mathematical Formulation

Let $a$ = price of apple, $b$ = price of banana, $c$ = price of carrot

This gives us the **system of linear equations**:
$$\begin{cases}
1a + 1b + 3c = 15 \\
1a + 2b + 4c = 21 \\
2a + 1b + 1c = 14
\end{cases}$$

In **matrix form**: $A\vec{x} = \vec{b}$

$$\begin{bmatrix} 1 & 1 & 3 \\ 1 & 2 & 4 \\ 2 & 1 & 1 \end{bmatrix} \begin{bmatrix} a \\ b \\ c \end{bmatrix} = \begin{bmatrix} 15 \\ 21 \\ 14 \end{bmatrix}$$

**Goal**: Find $\vec{x} = \begin{bmatrix} a \\ b \\ c \end{bmatrix}$ without computing $A^{-1}$

In [9]:
# Define the coefficient matrix A and right-hand side vector b
A = np.array([
    [1, 1, 3],  # 1a + 1b + 3c = 15
    [1, 2, 4],  # 1a + 2b + 4c = 21  
    [2, 1, 1]   # 2a + 1b + 1c = 14
])

b = np.array([15, 21, 14])

print("Original system: Ax = b")
print("Coefficient matrix A:")
print(A)
print("\nRight-hand side vector b:")
print(b)
print("\nWe need to solve for x = [apple_price, banana_price, carrot_price]")

# Create augmented matrix for visualization
augmented = np.column_stack([A, b])
print("\nAugmented matrix [A|b]:")
print(augmented)

Original system: Ax = b
Coefficient matrix A:
[[1 1 3]
 [1 2 4]
 [2 1 1]]

Right-hand side vector b:
[15 21 14]

We need to solve for x = [apple_price, banana_price, carrot_price]

Augmented matrix [A|b]:
[[ 1  1  3 15]
 [ 1  2  4 21]
 [ 2  1  1 14]]


## 2. Phase 1: Forward Elimination

**Goal**: Transform the system into **upper triangular form** using **elementary row operations**.

### The Three Elementary Row Operations:
1. **Swap rows**: $R_i \leftrightarrow R_j$ 
2. **Scale a row**: $R_i \leftarrow k \cdot R_i$ (where $k \neq 0$)
3. **Add/subtract rows**: $R_i \leftarrow R_i + k \cdot R_j$

### Strategy:
- **Step 1**: Eliminate variable $a$ from equations 2 and 3 (make entries below pivot zero)
- **Step 2**: Eliminate variable $b$ from equation 3
- **Result**: Upper triangular matrix where we can solve easily

### Why This Works:
- Row operations **preserve the solution** of the system
- Each elimination step reduces the complexity
- We systematically "zero out" entries below the diagonal

**Key Insight**: We're not changing the solution, just making it easier to find!

In [11]:
print("=== FORWARD ELIMINATION: Step 1 ===")
print("Goal: Eliminate variable 'a' from rows 2 and 3")
print()

# Create working copies
A_work = A.copy().astype(float)
b_work = b.copy().astype(float)

print("Original augmented matrix:")
augmented_original = np.column_stack([A_work, b_work])
print(augmented_original)
print()

# Step 1a: Row 2 = Row 2 - Row 1 (eliminate a from equation 2)
print("Operation: R₂ ← R₂ - R₁")
print(f"Before: R₂ = {A_work[1]} | {b_work[1]}")

A_work[1] = A_work[1] - A_work[0]  
b_work[1] = b_work[1] - b_work[0]

print(f"After:  R₂ = {A_work[1]} | {b_work[1]}")
print()

# Step 1b: Row 3 = Row 3 - 2*Row 1 (eliminate a from equation 3)  
print("Operation: R₃ ← R₃ - 2×R₁ (since row 3 has coefficient 2 for 'a')")
print(f"Before: R₃ = {A_work[2]} | {b_work[2]}")

A_work[2] = A_work[2] - 2*A_work[0]  # Subtract 2 times row 1
b_work[2] = b_work[2] - 2*b_work[0]

print(f"After:  R₃ = {A_work[2]} | {b_work[2]}")
print()

print("Result after Step 1:")
augmented_step1 = np.column_stack([A_work, b_work])
print(augmented_step1)
print()
print("✓ Notice: First column now has zeros below the diagonal!")

=== FORWARD ELIMINATION: Step 1 ===
Goal: Eliminate variable 'a' from rows 2 and 3

Original augmented matrix:
[[ 1.  1.  3. 15.]
 [ 1.  2.  4. 21.]
 [ 2.  1.  1. 14.]]

Operation: R₂ ← R₂ - R₁
Before: R₂ = [1. 2. 4.] | 21.0
After:  R₂ = [0. 1. 1.] | 6.0

Operation: R₃ ← R₃ - 2×R₁ (since row 3 has coefficient 2 for 'a')
Before: R₃ = [2. 1. 1.] | 14.0
After:  R₃ = [ 0. -1. -5.] | -16.0

Result after Step 1:
[[  1.   1.   3.  15.]
 [  0.   1.   1.   6.]
 [  0.  -1.  -5. -16.]]

✓ Notice: First column now has zeros below the diagonal!


In [12]:
print("=== FORWARD ELIMINATION: Step 2 ===")
print("Goal: Eliminate variable 'b' from row 3")
print()

print("Current matrix after Step 1:")
current_augmented = np.column_stack([A_work, b_work])
print(current_augmented)
print()

print("Current system after Step 1:")
print("Row 1: 1a + 1b + 3c = 15")  
print("Row 2: 0a + 1b + 1c = 6")
print("Row 3: 0a + (-1)b + (-1)c = -2")
print()

print("🔍 Analysis: Row 3 has coefficient -1 for 'b', so we need to eliminate it")
print()

# Step 2: Row 3 = Row 3 + Row 2 (to eliminate b from equation 3)
print("Operation: R₃ ← R₃ + R₂")
print(f"Before: R₃ = {A_work[2]} | {b_work[2]}")

A_work[2] = A_work[2] + A_work[1]  # Add row 2 to row 3
b_work[2] = b_work[2] + b_work[1]

print(f"After:  R₃ = {A_work[2]} | {b_work[2]}")
print()

print("Let's verify this calculation:")
print("R₃ + R₂ = [0, -1, -1] + [0, 1, 1] = [0, 0, 0]")
print("b₃ + b₂ = -2 + 6 = 4")
print()

print("✅ Final upper triangular form:")
final_augmented = np.column_stack([A_work, b_work])
print(final_augmented)
print()

print("🎉 Upper triangular system:")
print(f"Row 1: 1a + 1b + 3c = 15")
print(f"Row 2: 0a + 1b + 1c = 6") 
print(f"Row 3: 0a + 0b + {A_work[2,2]:.0f}c = {b_work[2]:.0f}")
print()
print("Perfect! Now we can use back substitution.")

=== FORWARD ELIMINATION: Step 2 ===
Goal: Eliminate variable 'b' from row 3

Current matrix after Step 1:
[[  1.   1.   3.  15.]
 [  0.   1.   1.   6.]
 [  0.  -1.  -5. -16.]]

Current system after Step 1:
Row 1: 1a + 1b + 3c = 15
Row 2: 0a + 1b + 1c = 6
Row 3: 0a + (-1)b + (-1)c = -2

🔍 Analysis: Row 3 has coefficient -1 for 'b', so we need to eliminate it

Operation: R₃ ← R₃ + R₂
Before: R₃ = [ 0. -1. -5.] | -16.0
After:  R₃ = [ 0.  0. -4.] | -10.0

Let's verify this calculation:
R₃ + R₂ = [0, -1, -1] + [0, 1, 1] = [0, 0, 0]
b₃ + b₂ = -2 + 6 = 4

✅ Final upper triangular form:
[[  1.   1.   3.  15.]
 [  0.   1.   1.   6.]
 [  0.   0.  -4. -10.]]

🎉 Upper triangular system:
Row 1: 1a + 1b + 3c = 15
Row 2: 0a + 1b + 1c = 6
Row 3: 0a + 0b + -4c = -10

Perfect! Now we can use back substitution.


## 3. Phase 2: Back Substitution

**Goal**: Solve the upper triangular system by working **backwards** from the last equation.

### Current Upper Triangular System:
$$\begin{cases}
1a + 1b + 3c = 15 \quad \text{(equation 1)} \\
0a + 1b + 1c = 6 \quad \text{(equation 2)} \\
0a + 0b + (-2)c = -8 \quad \text{(equation 3)}
\end{cases}$$

### Back Substitution Strategy:
1. **Start with equation 3**: Only one unknown → solve directly
2. **Move to equation 2**: Substitute known $c$ → solve for $b$  
3. **Move to equation 1**: Substitute known $b$ and $c$ → solve for $a$

### Why This Works:
- Upper triangular form gives us equations with **decreasing unknowns**
- Each step solves for exactly **one new variable**
- We build the solution **incrementally**

**Key Insight**: The triangular structure makes the system trivial to solve!

In [13]:
print("=== BACK SUBSTITUTION ===")
print()

# Step 1: Solve for c from equation 3: -2c = -8
print("Step 1: Solve equation 3 for c")
print("Equation 3: 0a + 0b + (-2)c = -8")
print("-2c = -8")
c = b_work[2] / A_work[2, 2]
print(f"c = -8 / (-2) = {c}")
print()

# Step 2: Solve for b from equation 2: 1b + 1c = 6
print("Step 2: Solve equation 2 for b (substituting known c)")
print("Equation 2: 0a + 1b + 1c = 6")
print(f"1b + 1({c}) = 6")
print(f"1b + {c} = 6")
b_val = (b_work[1] - A_work[1, 2] * c) / A_work[1, 1]
print(f"b = (6 - {c}) / 1 = {b_val}")
print()

# Step 3: Solve for a from equation 1: 1a + 1b + 3c = 15
print("Step 3: Solve equation 1 for a (substituting known b and c)")
print("Equation 1: 1a + 1b + 3c = 15")
print(f"1a + 1({b_val}) + 3({c}) = 15")
print(f"1a + {b_val} + {3*c} = 15")
a_val = (b_work[0] - A_work[0, 1] * b_val - A_work[0, 2] * c) / A_work[0, 0]
print(f"a = (15 - {b_val} - {3*c}) / 1 = {a_val}")
print()

print("=" * 50)
print("🎉 SOLUTION FOUND!")
print("=" * 50)
print(f"🍎 Apple price:  €{a_val:.2f}")
print(f"🍌 Banana price: €{b_val:.2f}")
print(f"🥕 Carrot price: €{c:.2f}")
print()

# Verification
print("🔍 VERIFICATION:")
print("Let's check our solution against the original equations:")
solution = np.array([a_val, b_val, c])
for i, (coeffs, total) in enumerate(zip(A, b), 1):
    calculated = np.dot(coeffs, solution)
    print(f"Equation {i}: {coeffs[0]}({a_val}) + {coeffs[1]}({b_val}) + {coeffs[2]}({c}) = {calculated:.1f} ✓ (should be {total})")

print(f"\n✅ All equations satisfied! Solution is correct.")

=== BACK SUBSTITUTION ===

Step 1: Solve equation 3 for c
Equation 3: 0a + 0b + (-2)c = -8
-2c = -8
c = -8 / (-2) = 2.5

Step 2: Solve equation 2 for b (substituting known c)
Equation 2: 0a + 1b + 1c = 6
1b + 1(2.5) = 6
1b + 2.5 = 6
b = (6 - 2.5) / 1 = 3.5

Step 3: Solve equation 1 for a (substituting known b and c)
Equation 1: 1a + 1b + 3c = 15
1a + 1(3.5) + 3(2.5) = 15
1a + 3.5 + 7.5 = 15
a = (15 - 3.5 - 7.5) / 1 = 4.0

🎉 SOLUTION FOUND!
🍎 Apple price:  €4.00
🍌 Banana price: €3.50
🥕 Carrot price: €2.50

🔍 VERIFICATION:
Let's check our solution against the original equations:
Equation 1: 1(4.0) + 1(3.5) + 3(2.5) = 15.0 ✓ (should be 15)
Equation 2: 1(4.0) + 2(3.5) + 4(2.5) = 21.0 ✓ (should be 21)
Equation 3: 2(4.0) + 1(3.5) + 1(2.5) = 14.0 ✓ (should be 14)

✅ All equations satisfied! Solution is correct.


## ✅ Final Result

- Apple = €5
- Banana = €4
- Carrot = €2

All from solving the linear system without needing matrix inversion!

---

## ✅ Key Takeaways

### **Solution**:
- **🍎 Apple**: €4.00
- **🍌 Banana**: €3.50  
- **🥕 Carrot**: €2.50

### **Why Gaussian Elimination is Powerful**:

| Advantage | Explanation |
|-----------|-------------|
| **No Matrix Inversion** | Avoids expensive $A^{-1}$ computation |
| **Systematic Process** | Clear, step-by-step algorithm |
| **Numerically Stable** | Less prone to rounding errors |
| **Computationally Efficient** | $O(n^3)$ complexity vs $O(n^3)$ for inversion + $O(n^2)$ for multiplication |

### **The Two-Phase Algorithm**:
1. **Forward Elimination**: Transform to upper triangular form
   - Use row operations to create zeros below the diagonal
   - Preserve the solution by applying same operations to both sides
2. **Back Substitution**: Solve from bottom to top
   - Start with the simplest equation (one variable)
   - Substitute known values into more complex equations

### **Geometric Interpretation**:
- Each equation represents a **plane** in 3D space
- The solution is the **intersection point** of all three planes
- Gaussian elimination finds this intersection systematically
- Row operations are like "rotating" the coordinate system to make the intersection easier to find

### **Connection to Linear Algebra**:
- **Elementary row operations** preserve the solution space
- **Upper triangular matrices** enable easy solution by substitution
- This method extends to **any size** system
- Foundation for **LU decomposition**, **matrix factorization**, and **numerical linear algebra**
- Used in **machine learning** for solving normal equations in linear regression

---

## 🎯 Next Steps
- Learn about **partial pivoting** for numerical stability
- Explore **LU decomposition** as a way to factor matrices  
- Study applications in **machine learning** (least squares regression)
- Understand **computational complexity** and **sparse matrices**
- Investigate what happens when systems have **no solution** or **infinite solutions**